cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java

```diff
@@ -235,9 +230,31 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
         copyRuntimeToken(locationFactory, oldOptions.getArguments(), tempDir, localizeResources);
       }
 
+      Map<String, String> extraSystemArgs = new HashMap<>(launchConfig.getExtraSystemArguments());
+
+      // Localize the program jar
+      Location programJarLocation = program.getJarLocation();
+      String programJarName = programJarLocation.getName();
+      if (oldOptions.getArguments().hasOption(ProgramOptionConstants.PROGRAM_JAR_HASH)) {
+        // if hash value for program.jar has been provided, we append it to filename.
+        String programJarHash = oldOptions.getArguments().getOption(ProgramOptionConstants.PROGRAM_JAR_HASH);
+        programJarName = programJarName.replace(".jar", String.format("_%s%s", programJarHash, ".jar"));
+
+        Set<String> cacheableFiles = new HashSet<>();
+        if (oldOptions.getArguments().asMap().containsKey(ProgramOptionConstants.CACHEABLE_FILES)) {
+          cacheableFiles =
+            new HashSet<>(GSON.fromJson(oldOptions.getArguments().asMap().get(ProgramOptionConstants.CACHEABLE_FILES),
+                                        new TypeToken<Set<String>>() {
+                                        }.getType()));
```

Comment:
nit: a bit easier to read to put this on one line `new TypeToken<Set<String>>() { }.getType()`

Reply:
Somehow, intellij formatter does this. Fixed.